Add  `--kubernetes` flag to make dapr invoke cmd support kubernetes

imneov opened the pull request:
# Description

Support the invoke cmd on kubernetes mode. 

The process is similar:
1. Get the Pod corresponding to the App
2. Complete the specific communication through the `SubResource("proxy")` of `RESTClient`
3. Direct access to App's `AppPort` port or access to daprd's `HttpPort` port

```

                                      +-----------------+
                                      |                 |
                     AppPort          | +-------------+ |
         +----------------------------> |    APP      | |
  +----------+                        | +-------------+ |
  | dapr cli |                        |                 |
  +----------+                        | +-------------+ |
         +----------------------------> |    Daprd    | |
                    HttpPort          | +-------------+ |
                                      |                 |
                                      |            pod  |
                                      +-----------------+

```

## Issue reference

We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.

Please reference the issue this PR will close: #_[[436](https://github.com/dapr/cli/issues/436)]_ #_[[848](https://github.com/dapr/cli/issues/848)]_

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Extended the documentation

tanvigour reviewed (commented):
Can we add test case for this?

imneov commented:
> Can we add test case for this?

Sorry, just saw the reply.

Thank you tanvigour for review. I wrote part of the test code. There may be problems with code style or edge cases. Please review.

Welcome to add test cases.

imneov commented:
There is a problem with using `proxy` to access the `HttpPort`, mainly because the address that daprd listens to is `127.0.0.1`, and the proxy cannot access this port.

So now our project(https://github.com/tkeel-io/cli/issues/74) uses `portforward` for `invoke`. The basic process is as follows:

1. Get the Pod corresponding to the App
2. Build a tunnel between the local and target pod through RESTClient's SubResource("portforward")
3. Access daprd's `HttpPort` port through the local tunnel port

This part of the implementation needs to modify `portforward.go` to implement random ports (https://github.com/dapr/cli/issues/863).

mukundansundar commented:
@imneov Can you signoff on your commits? https://github.com/dapr/cli/blob/master/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work

mukundansundar commented:
@imneov Can you fix conflicts in the PR?

imneov commented:
> @imneov Can you fix conflicts in the PR?

Yes, I fix it.

shubham1172 commented on cmd/invoke.go:
```diff
@@ -39,15 +40,18 @@ var (
 
-# Invoke a sample method on target app with GET Verb
+# Invoke a sample method on target app with in Kubernetes
+dapr invoke -k --app-id target --method sample --data '{"key":"value"}
```
`--data` has an unclosed single quote. Same for the above example in self-hosted mode.

imneov replied:
Thank you for review. I fixed it.

shubham1172 commented on cmd/invoke.go:
```diff
@@ -79,11 +82,22 @@ dapr invoke --unix-domain-socket --app-id target --method sample --verb GET
-			print.FailureStatusEvent(os.Stderr, err.Error())
-			os.Exit(1)
+		var response string
+		if kubernetesMode {
```
Given that both kubernetes mode and standalone mode have same error handling, can we follow DRY and do this instead?
```go
if kubernetesMode {
    response, err = kubernetes.Invoke(invokeAppID, invokeAppMethod, bytePayload, invokeVerb)
} else {
    client := standalone.NewClient()
    response, err = client.Invoke(invokeAppID, invokeAppMethod, bytePayload, invokeVerb, invokeSocket)
}

// now handle error and response
if err != nil {
    // handle error
}
```

imneov replied:
Thank you for review. I fixed it.

shubham1172 commented on cmd/invoke.go:
```diff
@@ -79,7 +83,13 @@ dapr invoke --unix-domain-socket --app-id target --method sample --verb GET
+		} else {
+			client := standalone.NewClient()
+			response, err = client.Invoke(invokeAppID, invokeAppMethod, bytePayload, invokeVerb, invokeSocket)
+		}
```
nit: line break

imneov replied:
How can I fix this?
`make lint` without any hints.

shubham1172 replied:
Oh you can simply add a new line after line 92. It's only for readability purpose.

imneov replied:
OK, I fixed it.

imneov commented on pkg/kubernetes/invoke.go:
```diff
@@ -0,0 +1,178 @@
+				Namespace: p.Namespace,
+			}
+			for i, arg := range c.Args {
+				if arg == "--app-port" {
```
I agree with this, but the style is the same as
https://github.com/dapr/cli/blob/4765d02e5ab4675985aa59cbc6d5ecd91ff14b68/pkg/kubernetes/list.go#L49-L55

shubham1172 commented on cmd/invoke.go:
```diff
@@ -39,15 +40,19 @@ var (
-# Invoke a sample method on target app with GET Verb
+# Invoke a sample method on target app with in Kubernetes
+dapr invoke -k --app-id target --method sample --data '{"key":"value"}'
+
```
nit: remove extra line.

shubham1172 replied:
nit: remove extra line.

mukundansundar commented:
@imneov Can you fix the DCO check error please?

imneov commented:
> @imneov Can you fix the DCO check error please?

Done!

mukundansundar commented:
@pravinpushkar Please review this PR

pravinpushkar commented:
> @pravinpushkar Please review this PR

Looking at it now.

pravinpushkar commented:
@imneov Please fix the small lint issue. PR looks fine.

pravinpushkar commented:
ping @imneov

imneov commented:
> @imneov Please fix the small lint issue. PR looks fine. We need to update this here also to be in sync - https://docs.dapr.io/reference/cli/dapr-invoke/

Do I need to submit a PR for docs?

shubham1172 commented:
> > > @imneov Please fix the small lint issue. PR looks fine. We need to update this here also to be in sync - https://docs.dapr.io/reference/cli/dapr-invoke/
> 
> @pravinpushkar Sorry I didn't find the lint error. what command do I need to run?

The lint error is available in this view https://github.com/dapr/cli/pull/862/files
![image](https://user-images.githubusercontent.com/22132944/189104288-2e676618-74fe-441b-9d2d-6a0d4ca48ca7.png)

(there is only one of such).

imneov commented:
@pravinpushkar HI~  Is there anything else I need to do?

pravinpushkar commented:
> > @imneov Please fix the small lint issue. PR looks fine. We need to update this here also to be in sync - https://docs.dapr.io/reference/cli/dapr-invoke/
> 
> Do I need to submit a PR for docs?

Yes, If not done already then please create a doc issue and PR and link it here.
I have triggered the PR checks let's see.

imneov commented:
> Support the invoke cmd on kubernetes mode.

OK , Thanks for you reply.   @pravinpushkar 

There is  doc issue(https://github.com/dapr/docs/issues/2846) and PR(https://github.com/dapr/docs/pull/2847)

imneov commented:
@pravinpushkar Hi~ Is there any work I need to do?

mukundansundar reviewed (commented):
please add E2E for this feature.

msfussell commented:
@mukundansundar - What is happening with this issue/PR?

cicoyle commented:
@imneov - looks like there are some e2e failures, mind looking into and fixing those?
@dapr/maintainers-cli - looks like this PR has been open quite a while, we could use a review on this please :)

That being said, given the time constraint for this release, I will push this to the next release

imneov committed:
feat: support invoke cmd on kubernetes mode

Add `-k/--kubernetes` flag to `dapr invoke` so the command works against
apps running on a Kubernetes cluster.

The implementation locates the Pod hosting the target Dapr app, then uses
the API server `proxy` subresource on the pod's `appPort` to relay the
request, mirroring the existing `dapr list -k` pod discovery path.

README is updated with usage for both self-hosted and Kubernetes modes.

Closes #436, #848

Signed-off-by: imneov <i@neov.im>

imneov committed:
test: add kubernetes e2e for invoke -k

Deploys a minimal daprized nginx pod to a dedicated namespace, waits for
the daprd sidecar to be injected and the app to report Ready, and then
exercises `dapr invoke -k` via GET on the app. Also covers the negative
case where the target app-id does not exist.

Mirrors the TestCase helpers used by the existing K8s install suite so
install/uninstall runs in the same test as a setup/teardown pair.

Signed-off-by: imneov <i@neov.im>

imneov committed:
fix: satisfy staticcheck QF1003 and move e2e teardown out of Cleanup

- pkg/kubernetes/invoke.go: convert the else-if chain in getAppInfoFromPod
  to a tagged switch. staticcheck QF1003 is a hard failure under
  golangci-lint v2.10.1.
- tests/e2e/kubernetes/invoke_test.go: move uninstall subtests out of
  t.Cleanup into a deferred closure. t.Run is not valid from t.Cleanup
  because the test is already finalized, but it is valid from a defer
  since defers run before the test function returns.

Signed-off-by: imneov <i@neov.im>

imneov commented:
Rebased onto current master as a single clean commit, added a dedicated kubernetes E2E and fixed two real CI blockers on the latest run:

1. `pkg/kubernetes/invoke.go` — converted the `else-if` chain in `getAppInfoFromPod` to a tagged `switch`. `staticcheck QF1003` is a hard failure under `golangci-lint v2.10.1`, which is what was breaking `Build linux_amd64 binaries`. This is the lint @pravinpushkar flagged a while back.

2. `tests/e2e/kubernetes/invoke_test.go` (new) — end-to-end coverage for `dapr invoke -k`: deploys a daprized nginx pod, waits for the sidecar, runs `dapr invoke -k --method / --verb GET` and asserts the response. Also covers the "app not found" negative path. The first run actually passed all four body subtests; the parent failed because I was calling `t.Run` from inside `t.Cleanup`, which Go forbids — moved teardown into a `defer` where `t.Run` is allowed.

The remaining failing jobs (`Upgrade path tests` HA, `Self-hosted darwin`) are flaky on master itself — last 20 scheduled runs on master show ~50% and ~35% failure rates respectively, with success/failure alternating minutes apart. They need a rerun rather than a fix.

Could one of @cicoyle / @pravinpushkar / @mukundansundar take another look when you have a moment? The docs counterpart (dapr/docs#2847) was auto-closed while this sat — I'll re-open a fresh one as soon as this merges.
